tests/reframe/eessi-checks/applications/gromacs.py

```diff
@@ -0,0 +1,78 @@
+import re
+import reframe as rfm
+from reframe.utility import find_modules
+
+from testlib.applications.gromacs import Gromacs
+
+@rfm.required_version('>=3.5.0')
+@rfm.simple_test
+class Gromacs_EESSI(Gromacs):
+    '''EESSI Gromacs check.
+    This test will run GROMACS using all modules with 'GROMACS' in the module environment it can find.
+    On GPU nodes, it will only run tests if module names also contain 'cuda'.
+    On CPU nodes, it will only run tests if a module name does NOT contain 'cuda'.
+    Whether a nodes is CPU/GPU is determined based on if a device named 'gpu' is specified in the ReFrame settings file for the current partition.
+    Number of tasks, tasks per node and cpus per task are set based on the number of GPUs and number of CPUs specified in the ReFrame config file for the current partition. 
+    '''
+
+    scale = parameter(['singlenode', 'small', 'large'])
+    module_info = parameter(find_modules('GROMACS', environ_mapping={r'.*': 'builtin'}))
+
+    @rfm.run_after('init')
+    def apply_module_info(self):
+        self.s, self.e, self.m = self.module_info
+        self.valid_systems = [self.s]
+        self.modules = [self.m]
+        self.valid_prog_environs = [self.e]
+
+    @rfm.run_after('init')
+    def set_test_scale(self):
+        if self.scale == 'singlenode':
+            self.nsteps = 10000
+            self.num_nodes = 1
+        elif self.scale == 'small':
+            self.nsteps = 40000
+            self.num_nodes = 4
+        elif self.scale == 'large':
+            self.nsteps = 100000
+            self.num_nodes = 10
+        self.tags.add(self.scale)
+
+    @rfm.run_after('init')
+    def requires_gpu(self):
+        self.requires_cuda = False
+        if re.search("(?i)cuda", self.m) is not None:
+            self.requires_cuda = True
+
+    @rfm.run_after('setup')
+    def check_gpu_presence(self):
+        self.gpu_list = [ dev.num_devices for dev in self.current_partition.devices if dev.device_type == 'gpu' ]
```

Comment:
It feels to me that you might be using this hook in many other tests too. If so, you can move this `check_gpu_presence` function into a hook utility module and then bind the external function into the class as a hook. To bind it, you can either do
```python
import my_hook_utils as hu

class MyTest(rfm.RunOnlyTest):
    ...
    @run_after('setup')
    def check_gpu_presence(self):
        hu.check_gpu_presence(self)
```
or as a one-liner
```python
import my_hook_utils as hu

class MyTest(rfm.RunOnlyTest):
    ...
    run_after('setup')(bind(hu.check_gpu_presence))
```

Reply:
Oh, I like this idea! And it gives me another idea to: I can do the same thing for the requires_cuda test. The regex makes some pretty strong assumption (namely that `cuda` is in the name), but that only works for EB toolchains like `fosscuda` and `intelcuda`. It would fail e.g. with `gompic`. Moving it out of the test and into a `utils` kind of thing allows us to easily improve and/or adjust that check in a single place if things change in the future.